core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/ContainerTorrentDownloadJob.kt

```diff
@@ -0,0 +1,218 @@
+        val containerUid = uri.uri.toString().substringAfterLast("/").toLongOrNull() ?: return null
+
+        val container = repo.containerDao.findByUid(containerUid) ?: return null
+
+        val contentEntry = repo.contentEntryDao.findByUid(container.containerContentEntryUid)
+                ?: throw IllegalArgumentException("no entry found from container")
+
+        return MetadataResult(contentEntry as ContentEntryWithLanguage, PLUGIN_ID)
+    }
+
+    override suspend fun processJob(jobItem: ContentJobItemAndContentJob, process: ContentJobProcessContext, progress: ContentJobProgressListener): ProcessResult {
+        val contentJobItem = jobItem.contentJobItem ?: throw IllegalArgumentException("missing job item")
+        val containerUid = contentJobItem.cjiContainerUid
+
+        // check if torrent file already exists
+        val torrentFile = File(torrentDir, "$containerUid.torrent")
+        return withContext(Dispatchers.Default) {
+
+            try {
+
+                if (!torrentFile.exists()) {
+                    val path = UMFileUtil.joinPaths(endpoint.url, "containers/$containerUid")
+                    val torrentFileStream = httpClient.get<InputStream>(path)
+                    torrentFile.writeBytes(torrentFileStream.readBytes())
+                }
+
+                val downloadFolderPath = jobItem.contentJob?.toUri?.let { URI(it).path }
+                        ?: containerDir.path
+
+                val containerFilesFolder = File(downloadFolderPath, containerUid.toString())
+                containerFilesFolder.mkdirs()
+
+                val metadataProvider = FileMetadataProvider(torrentFile.absolutePath)
+                val fileNameList = metadataProvider
+                        .torrentMetadata
+                        .files
+                        .map { File(it.relativePathAsString).name }
+
+                val existingFiles = db.containerEntryFileDao.findEntriesByMd5SumsSafeAsync(fileNameList,
+                        db.maxQueryParamListSize)
+
+                //link them to the new container path
+                existingFiles.forEach {
+                    val existingPath = it.cefPath ?: return@forEach
+                    val existingFile = File(existingPath)
+                    if (existingFile.name == MANIFEST_FILE_NAME) {
+                        return@forEach
+                    }
+                    val target = File(containerFilesFolder, existingFile.name)
+                    if (!target.exists()) {
+                        createSymLink(existingPath, target.path)
+                    }
+                }
+
+                withWifiLock(context) {
+
+                    val startTime = System.currentTimeMillis()
+                    ustadTorrentManager.addTorrent(containerUid, containerFilesFolder.path)
+
+                    val torrentDeferred = CompletableDeferred<Boolean>()
+                    val torrentListener = object : TorrentDownloadListener {
```

Comment:
This listener does not handle the failure case! Right now, in case of failure, it will get stuck forever. When the failure event is received, that is the time to call torrentDeferred.completeExceptionally.